File: tests/config.test.mjs
import { describe, expect, it } from 'vitest';
import { mkdtemp, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import { loadConfig, normalizeConfig } from '../src/config.mjs';

async function tmpConfig(body) {
  const dir = await mkdtemp(path.join(tmpdir(), 'theme-config-'));
  const file = path.join(dir, 'themes.yml');
  await writeFile(file, body);
  return file;
}

const minimal = `
discourse:
  repo_path: /tmp/discourse
  base_url: https://example.com
themes:
  - id: test-theme
    name: Test Theme
    source:
      type: git
      url: https://github.com/example/test-theme.git
`;

describe('config', () => {
  it('loads the example config', async () => {
    const config = await loadConfig('config/themes.example.yml', { checkRepoPath: false });
    expect(config.themes.map((t) => t.id)).toEqual([
      'discourse-verso',
      'discourse-air',
      'discourse-mint-theme',
      'minima',
    ]);
    expect(config.screenshot.width).toBe(2560);
  });

  it('rejects missing theme id', async () => {
    const file = await tmpConfig(minimal.replace('id: test-theme\n    ', ''));
    await expect(loadConfig(file, { checkRepoPath: false })).rejects.toThrow(/theme 1.*id/i);
  });

  it('rejects unknown source type', async () => {
    const file = await tmpConfig(minimal.replace('type: git', 'type: zip'));
    await expect(loadConfig(file, { checkRepoPath: false })).rejects.toThrow(/source.type/i);
  });

  it('applies screenshot defaults and modes', () => {
    const config = normalizeConfig({
      discourse: { repo_path: '/tmp/discourse', base_url: 'https://example.com' },
      themes: [{ id: 'x', name: 'X', source: { type: 'local', path: '/tmp/x' } }],
    }, { checkRepoPath: false });
    expect(config.screenshot).toMatchObject({ width: 2560, height: 1440, format: 'webp', max_bytes: 1048576 });
    expect(config.themes[0].modes).toEqual(['light', 'dark']);
  });
});

File: test/config_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "tmpdir"
require_relative "../lib/theme_screenshots/config"

class ConfigTest < Minitest::Test
  def test_loads_theme_urls_without_requiring_local_clones
    Dir.mktmpdir do |dir|
      path = File.join(dir, "themes.yml")
      File.write(path, <<~YAML)
        discourse:
          repo_path: /tmp/discourse
        themes:
          - id: minima
            name: Minima
            source:
              type: git
              url: https://github.com/Discourse/minima.git
      YAML

      config = ThemeScreenshots::ConfigLoader.load(path)
      assert_equal "/tmp/discourse", config.discourse["repo_path"]
      assert_equal "https://github.com/Discourse/minima.git", config.themes.first.theme_url
      assert_equal %w[light], config.themes.first.modes
      assert_equal "topic", config.screenshot["subset"]
    end
  end

  def test_loads_local_theme_paths
    Dir.mktmpdir do |dir|
      path = File.join(dir, "themes.yml")
      File.write(path, <<~YAML)
        discourse:
          repo_path: /tmp/discourse
        themes:
          - id: zleek
            name: zleek
            source:
              type: local
              path: /tmp/discourse-zleek
      YAML

      config = ThemeScreenshots::ConfigLoader.load(path)
      assert_equal "local", config.themes.first.source["type"]
      assert_equal "/tmp/discourse-zleek", config.themes.first.theme_url
      assert_equal "local:/tmp/discourse-zleek", config.themes.first.source_label
    end
  end

  def test_loads_core_theme_entries_for_github_actions
    Dir.mktmpdir do |dir|
      path = File.join(dir, "themes.yml")
      File.write(path, <<~YAML)
        discourse:
          repo_path: ../discourse
        themes:
          - id: foundation
            name: Foundation
            source:
              type: core
      YAML

      config = ThemeScreenshots::ConfigLoader.load(path)
      assert_equal "core", config.themes.first.source["type"]
      assert_equal "foundation", config.themes.first.theme_url
      assert_equal "core:foundation", config.themes.first.source_label
    end
  end
end
